IMOD/manpages/endmodel.man
.na
.nh
.TH endmodel 1 4.6.34 IMOD
.SH NAME
endmodel - make a model file with starting & ending points of contours
.SH SYNOPSIS
endmodel
.SH DESCRIPTION
Endmodel will make a new model file containing only specified
starting and ending points of contours in a model.  The contours in the
the original objects may end up in completely different new model object.
.P
Entries to the program:
.P
Name of input model file
.P
Name of output model file
.P
Number of different objects to create in the output
model.  For each such object, next enter the following two lines:
.P
   "Color" of the object in the output model, and number of kinds of
   ends to include in that object.  This "color" is 256 minus the
   desired object number.  (For example, "250,4" to include
   four kinds of ends in object 6)
.P
   A list of pairs of values: each pair consists of an object number
   in the old model, and either a 0 to take the starting point or a 1
   to take the ending point for each contour in that object.  (For
   example, to take starting points of contours in objects 1 and 3,
   and ending points of contours in objects 2 and 4, enter the list
   "1,0,3,0,2,1,4,1" )
.P
Number of points to generate for each new model contour, and the
increment in Z between them.  If you need this model just to look at,
enter "1,0" to get just a single point.  If you intend to run an MTK
analysis of 3-D distances between these ends, you need at least 2
points; in this case enter "2,0.1"
.SH HISTORY
.nf
Written by David Mastronarde, 11/23/92;  modified for IMOD 4/24/97
.fi
.SH BUGS
Email bug reports to mast@colorado.edu.


IMOD/History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
10/8/23: Fixed crash in Ctfphaseflip with X size of 4092 (i.e., K3 camera) and
tilt angle of exactly 0; fixed strip width at exactly 0 tilt with a maximum
width entered (In 4.11)

9/22/23: In Etomo - Fidual Model, fixed the "Use Adjusted Track Com File"
button, which was not functioning (In 4.11).(Bug# 2458)

3/14/23: Fixed Batchruntomo problem running Splitcorrection with Python 3
(In 4.11).

2/3/23: IMOD_4-11-24
2/3/23: Fixed Autofidseed problem with Python 3 (In 4.11).

1/10/22: IMOD_4-11-23
1/10/22: Fixed Sampletilt when the dataset name contains "slice" (In 4.11).

1/8/22: IMOD_4-11-22
12/22/22: Fixed a problem where most of the interfaces are unable to handle a
missing file name style (In 4.11).(Bug# 2427)

12/21/22: Fixed initialization of the back projection checkbox in batch
interface (in 4.11).(Bug# 2417)

12/7/22: Fixed Tiltalign crash from array that was too small when there are
many views and few points (In 4.11, already fixed in 4.12).

11/1/22: IMOD_4-11-21
11/1/22: Fixed failure in Copytomocoms from not converting some directive
values to float (In 4.11).

9/27/22: Add entry for autofidseed LowerTargetForClustered option.(Bug# 2418)

9/27/22: In Etomo - Batchruntomo Interface, fixed a problem with the
backprojection checkbox (In 4.11).(Bug# 2417)

9/7/22: Changes to framewatcher to watch for EER files (In 4.11).

9/2/22: IMOD_4-11-20
9/1/22: In Etomo - Batchruntomo Interface, fixed some problems with the back-
projection checkbox (In 4.11).(Bug# 2417)

8/25/22: Switched to Qt 5.15.12 for Linux builds to solve a problem with being
unable to resize or move windows after mouse-rotating in slicer (In 4.11).

8/3/22: Fixed 3dmod crash if model loaded during initial image load (In 4.11).

7/29/22: IMOD_4-11-19
7/29/22: Fixed no new line before MESSAGE: from Tilt if GPU not available
(In 4.11).

7/28/22: In Etomo - PEET, fixed a problem where the processchunks monitor did
not recognize any single-line messages from processchunks.out (In 4.11).
(Bug# 2321)

7/18/22: Changed defocus tolerance in ctfcorrection command file to 50 and
made Ctf3dsetup and Subtomosetup use the slab thickness for this tolerance, 
since it determines the width of regions near the edges corrected at just one
defocus (In 4.11).

7/12/22 In Etomo - Final Aligned Stack handling a config file with an
incorrect path (In 4.11).(Bug# 2404)

7/8/22 In Etomo - Tomogram Generation fixed the choosing of SIRT output
tomograms to load into 3dmod (In 4.11).(Bug# 2402)

7/7/22 In Etomo - Reconstruction Interface added missing image file name style
recovery (In 4.11).(Bug# 2403)

7/1/22: Re-enabled use of RGB data in Drawing Tools (In 4.11).

6/30/22: In Etomo - Batchruntomo Interface fixed a problem where a delivery
was done on every run (In 4.11).((Bug# 2405)

6/22/22: The minimum zero spacing in CTF correction is now defined as a float
and setting it will no longer generate an error in Etomo (In 4.11).(Bug# 2406)

6/18/22: IMOD_4-11-18
6/17/22: Fixed summing into power spectrum in Ctfplotter, which was off by a
pixel in the upper half of the FFT and biased defocus by 10-20 nm (In 4.11).

6/15/22: IMOD_4-11-17
6/14/22: Fixed wrong scaling of spherical aberration in Ctfphaseflip, which
made it correct effectively 20-50 nm higher defocus and get off above ~5 A
resolution (In 4.11).

6/11/22: Fixed Ctfplotter with pixel cropping to take tiles from correct
locations; it was sampling more from one side than the other, not adjusting
each spectrum properly for tile location, blurring the power spectrum, and
giving increasingly biased defocus estimates with increasing tilt (In 4.11).

6/10/22: In Etomo - Reconstruction interface, fixing a problem that prevented
a dataset created by batchruntomo from being loaded (In 4.11).(Bug# 2409)

6/8/22: Ctfplotter: fixed the crop pixel option when running
non-interactively; it was either not cropping or going into an infinite
loop. Fixed the starting point for autofitting to be near zero tilt 
when running non-interactively.  Fixed a memory leak (In 4.11).

6/6/22: Fixed Zap and Slicer snapshots having rubberband on edge with high-DPI
monitors (In 4.11).

6/2/22: Fixed crash in Ctfphaseflip with GPU from accessing strip coordinates
out of range (In 4.11).

5/30/22: IMOD_4-11-16
5/30/22: Made sure autocontrasting does not happen until the Zap window has
been drawn (In 4.11).

5/26/22: Made Batchruntomo detect and recover from an Etomo error in dataset
location when restarting from a step (In 4.11).

5/26/22: Fixed tile cache display in Zap window sometimes not being refreshed
after all the tiles are loaded (In 4.11).

5/25/22: Made Makepyramid work with a single-image file (In 4.11).

5/22/22: Changed installer for Linux/Mac to put link to python, if needed, in
a separate directory that is put on the back of the path (In 4.11).

5/20/22: Fixed skipping of global clip planes in model view (In 4.11).

5/20/22: Fixed sychronization of ghost and model view image display checkboxes
when toggling with hot keys (In 4.11).

5/20/22: Stopping limiting EER Z-summing to 99 in 3dmod (In 4.11).

5/6/22: Fixed a test for extension types in Splitblend, Splittilt, and
Splitcorrection (in 4.11).

4/26/22: Fixed direction of Step Up/Down buttons in Ctfplotter, broken 
1/22/20, and improved treatment of autofitting range (In 4.11).

4/19/22: Stopped ctfplotter from interpreting a zero astigmatism as not being
measured at all (In 4.11).

3/29/22: IMOD_4-11-15
3/29/22: Fixed Alignframes making unweighted sum with CPU (In 4.11).

3/21/22: IMOD_4-11-14
3/21/22: Changed installer for Linux to make link to python3 or python2 in
IMOD/bin instead of in /usr/bin when there is no "python" on the path, but for
Mac OS 12 it will look for python3 from python.org, homebrew, or Apple Developer
Tools and make the link directly to the first one it finds (In 4.11). 


IMOD/flib/model/endmodel.f
*       * * * * ENDMODEL * * * * *
c       
c       ENDMODEL will make a new model file containing only specified
c       starting and ending points of contours in a model.  The colors of
c       the new model objects may be completely different from the colors of
c       the original objects.
c       
c       For data coming from a WIMP model file, the IMOD object number is
c       the 256 minus the WIMP color (e.g., object 1 for color 255).  WIMP
c       objects that are turned off may be referred to by the negative of
c       this object number.
c       
c       Entries to the program:
c       
c       Name of input model file
c       
c       Name of output model file
c       
c       Number of different objects to create in the output
c       model.  For each such object, next enter the following two lines:
c       
c       .  Color of the object in the output model, and number of kinds of
c       .  ends to include in that object.  (For example, "250,4" to include
c       .  four kinds of ends in an object colored 250)
c       
c       .  A list of pairs of values: each pair consists of an object number
c       .  in the old model, and either a 0 to take the starting point or a 1
c       .  to take the ending point for each contour in that object.  (For
c       .  example, to take starting points of contours in objects 1 and 3,
c       .  and ending points of contours in objects 2 and 4, enter the list
c       .  "1,0,3,0,2,1,4,1" )
c       
c       Number of points to generate for each new model contour, and the
c       increment in Z between them.  If you need this model just to look at,
c       enter "1,0" to get just a single point.  If you intend to run an MTK
c       analysis of 3-D distances between these ends, you need at least 2
c       points; in this case enter "2,0.1"
c       
c       David Mastronarde  11/23/92; modified for IMOD 4/24/97
c       
      use fortmodel
      character*320 modelfile
      logical exist,readw_or_imod
      real*4, allocatable :: xyzmt(:,:,:)
      integer*4, allocatable :: itype(:)
      integer*4 newtype(100),nmap(100) ,ioldtype(100,100),ifend(100,100)
c       
      call setExitPrefix('ERROR: ENDMODEL - ')
91    write(*,'(1x,a,$)')'Name of input model file: '
      read(5,'(a)')modelfile
c       
75    exist=readw_or_imod(modelfile)
      if(.not.exist)go to 91
      allocate(itype(max_obj_num), xyzmt(3,2,max_obj_num), stat = ierr)
      call memoryError(ierr, 'arrays for objects')

c       
      write(*,'(1x,a,$)')'Name of model file to store points in: '
      read(5,'(a)')modelfile
c       
      call deleteiobj()
      nmt=0
      do iobj=1,max_mod_obj
        if(npt_in_obj(iobj).ge.2)then
          ibase=ibase_obj(iobj)
          ip1=abs(object(1+ibase))
          ip2=abs(object(npt_in_obj(iobj)+ibase))
          nmt=nmt+1
          itype(nmt)=256-obj_color(2,iobj)
          if(obj_color(1,iobj).eq.0)itype(nmt)=-itype(nmt)
          do i=1,3
            xyzmt(i,1,nmt)=p_coord(i,ip1)
            xyzmt(i,2,nmt)=p_coord(i,ip2)
          enddo
        endif
      enddo
c       
      write(*,'(1x,a,$)')'# of new objects to create: '
      read(5,*)nnew
c       
      do inew=1,nnew
        write(*,'(1x,a,i3,a,$)')'For new object #',inew,
     &      ', enter new color and # of objects mapping to it: '
        read(5,*)newtype(inew),nmap(inew)
        print *,'For each old object, enter the object # (- if ',
     &      'turned off in WIMP model),',
     &      ' and 0 for start or 1 for end'
        read(5,*)(ioldtype(inew,j),ifend(inew,j),j=1,nmap(inew))
      enddo
c       
      write(*,'(1x,a,$)')
     &    '# of points to store at each end, z increment: '
      read(5,*)nstore,zinc
c       
      do i=1,max_obj_num
        npt_in_obj(i)=0
      enddo
      n_object=0
      n_point=0
c       
      do imt=1,nmt
        do inew=1,nnew
          do imap=1,nmap(inew)
            if(itype(imt).eq.ioldtype(inew,imap))then
              n_object=n_object+1
              obj_color(1,n_object)=1
              obj_color(2,n_object)=newtype(inew)
              npt_in_obj(n_object)=nstore
              ibase_obj(n_object)=n_point
              isn=ifend(inew,imap)+1
              do inc=1,nstore
                n_point=n_point+1
                object(n_point)=n_point
                p_coord(1,n_point)=xyzmt(1,isn,imt)
                p_coord(2,n_point)=xyzmt(2,isn,imt)
                p_coord(3,n_point)=xyzmt(3,isn,imt)+(inc-1)*zinc
                pt_label(n_point)=0
              enddo
            endif
          enddo
        enddo
      enddo
c       
      max_mod_obj=n_object
      call write_wmod(modelfile)
      stop
      end
